yage-glutin/src/application.rs

```diff
@@ -0,0 +1,176 @@
+    ///
+    /// - `id`: Window ID
+    ///
+    /// # Returns
+    ///
+    /// Mutable reference to the window.
+    ///
+    pub fn window(&mut self, id: glutin::WindowId) -> Option<&mut Window> {
+        self.windows.get_mut(&id)
+    }
+
+    ///
+    /// Borrow events loop
+    ///
+    /// # Returns
+    ///
+    /// Reference to the events loop.
+    ///
+    pub fn events_loop(&self) -> &glutin::EventsLoop {
+        &self.events_loop
+    }
+
+    ///
+    /// Check if events loop is still running
+    ///
+    /// # Returns
+    ///
+    /// State of the events loop.
+    ///
+    pub fn is_running(&self) -> bool {
+        self.running
+    }
+
+    ///
+    /// Stop events loop
+    ///
+    /// This will stop the events loop and thereby exit the application.
+    ///
+    pub fn stop(&mut self) {
+        self.running = false;
+    }
+
+    ///
+    /// Poll events for all windows once.
+    ///
+    pub fn poll_events(&mut self) {
+        // get references to data we want to access, because closure borrows self
+        let events_loop = &mut self.events_loop;
+        let windows     = &self.windows;
+        let running     = &mut self.running;
+
+        // poll events
+        events_loop.poll_events(|event| {
+            // dispatch event
+            #[allow(clippy::single_match)]
+            match event {
+                // window events
+                glutin::Event::WindowEvent { event, window_id } => {
+                    // get window
+                    let window = windows.get(&window_id).unwrap();
+                    let gl_window = window.get_gl_window();
+
+                    // dispatch window event
+                    match event {
+                        // window closed
+                        glutin::WindowEvent::CloseRequested => {
+                            *running = false;
+                        },
+
+                        // window resized
+                        glutin::WindowEvent::Resized(logical_size) => {
+                            let dpi_factor = gl_window.get_hidpi_factor();
+                            gl_window.resize(logical_size.to_physical(dpi_factor));
+                        },
+
+                        // other event
+                        _ => (),
+                    }
+                },
+
+                // other event
+                _ => ()
+            }
+        });
+    }
+
+    ///
+    /// Run events loop
+    ///
+    /// Executes the events loop for the application.
+    /// This function will block and run as long as the events loop
+    /// is running. To stop the message loop, stop() has to be called
+    /// on the Application object.
+    ///
+    /// # Returns
+    ///
+    /// Exit code (0 for no error, > 0 for error)
+    ///
+    pub fn run(&mut self) -> i32 {
+        // run events loop until application is exited
+        while self.running {
+            self.poll_events();
+        }
+
+        // return value
+        0
```

Comment:
Numeric return codes are not very rusty (unless a process exit code is required). Prefer an enum (e.g. https://docs.rs/glutin/0.21.0-rc1/glutin/enum.ControlFlow.html).

Reply:
This actually IS the exit code of the application, I just didn't implement it fully at first.

Reply:
Ok, so this is more of a convenience for examples, right? Real applications would probably manage this themselves.

Reply:
No, real application should also use Application. That's what its there for. They can use Application::exit() to set the exit code when needed and use Application::run() to run the main loop. They should not implement it themselves, because it will handle the more complicated parts of a main loop for them.